### DISCUSSION TITLE
How to convert a .ok() match call to a pure match call when returning to a PyResult?

### DISCUSSION THREAD

**Discussion by nth10sd** [PyO3/pyo3#3486] [Questions] [answered]
```
use pyo3::exceptions::PyRuntimeError;
use pyo3::prelude::{pyfunction, PyResult};
use std::process::Command;
use which::which;

#[pyfunction]
pub fn list_func(program_name: &str) -> PyResult<bool> {
    match which(program_name).ok() {
        Some(ls_bin) => {
            let output = Command::new(ls_bin).arg("-l").output()?;
            if output.status.success() {
                Ok(true)
            } else {
                Ok(false)
            }
        }
        None => Err(PyRuntimeError::new_err("ls was not found on PATH")),
    }
}
```

In this example, the `which` call is fallible because of the `.ok()` usage. It should work with the `ls` program to list files, hence `ls -l`, but if the program is not found, the `which` call will fail.

May I know how I can drop the usage of the `.ok()`, making the errors that `which` returns go to the `Err` branch instead, whilst continuing to make the result happy with `PyResult<bool>`? Right now, if the input program does not exist, the code panics.

**Comment by birkenfeld** [comment, marked as answer]
`.ok()` does not introduce a panic. It merely converts a `Err(e)` which carries more information about the error into `None`, which just carries the information "it didn't work".
